File: src/PS3/FarmingMars.java
import java.util.HashMap;
import java.util.Scanner;

public class FarmingMars {

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int m = sc.nextInt();
		
		double[] farmland = new double[n];
		for (int i = 0; i < n; i++) {
			farmland[i] = sc.nextDouble();
		}
		
		boolean majorityFound = false;
		for (int i = 0; i < m; i++) {
			majorityFound = false;
			int l = sc.nextInt();
			int r = sc.nextInt();
			if (l == r) {
				System.out.println("usable");
				majorityFound = true;
				continue;
			}
			HashMap<Double, Integer> hm = new HashMap<Double, Integer>();
			int maj = (r - l + 1)/2 + 1;
			for (int j = l-1; j < r; j++) {
				if (hm.containsKey(farmland[j])) {
					int count = hm.get(farmland[j]) + 1;
					if (count >= maj) {
						System.out.println("usable");
						majorityFound = true;
					} else {
						hm.put(farmland[j], count);
					}
				} else {
					hm.put(farmland[j], 1);
				}
			}
			if (!majorityFound) {
				System.out.println("unusable");
			}
		}
	}

}

File: src/PS3/FarmingMars.cpp
#include <iostream>
#include <vector>
#include <unordered_map> 

using namespace std;

int main()
{
    int n = 0, m = 0, l = 0, r = 0;
    double k = 0;
    cin >> n;
    cin >> m;
    vector<double> farmland;
    bool majorityFound;

    for (int i = 0; i < n; i++) {
        cin >> k;
        farmland.push_back(k);
    }

    for (int i = 0; i < m; i++) {
        majorityFound = false;
        cin >> l;
        cin >> r;
        if (l == r) {
            cout << "usable" << endl;
            majorityFound = true;
        }
        else {
            unordered_map<double, int> umap;
            int maj = (r - l + 1) / 2 + 1;
            for (int j = l - 1; j < r; j++) {
                if (umap.find(farmland[j]) != umap.end()) {
                    int count = umap[farmland[j]] + 1;
                    if (count >= maj) {
                        cout << "usable" << endl;
                        majorityFound = true;
                    }
                    else {
                        umap[farmland[j]] = count;
                    }
                }
                else {
                    umap[farmland[j]] = 1;
                }
            }
            if (!majorityFound) {
                cout << "unusable" << endl;
            }
        }

    }

    return 0;
}